test/gpu/vllm/vllm_test.go

```diff
@@ -0,0 +1,193 @@
+		defer clientCtr.CleanUp(ctx)
+
+		// store vllm logs here
+		logsDir := b.TempDir()
+
+		b.StartTimer()
+		out, err := clientCtr.Run(ctx, dockerutil.RunOpts{
+			Links: []string{serverCtr.MakeLink("vllmctr")},
+			Image: "gpu/vllm",
+			Env:   []string{"PYTHONPATH=$PYTHONPATH:/vllm"},
+			Mounts: []mount.Mount{
+				// The logs dir is used because vllm only outputs json to a file.
+				{
+					Source: logsDir,
+					Target: "/tmp",
+					Type:   "bind",
+				},
+			},
+		}, "/vllm/benchmarks/benchmark_serving.py", "--num-prompts", fmt.Sprintf("%d", b.N), "--host", "vllmctr", "--model", "/model", "--tokenizer", "/model", "--endpoint", "/v1/completions", "--backend", "openai", "--dataset", "/ShareGPT_V3_unfiltered_cleaned_split.json", "--save-result", "--result-dir", "/tmp")
+		if err != nil {
+			b.Errorf("failed to run container: %v logs: %s", err, out)
+		}
+
+		b.StopTimer()
+
+		metrics, err := parseVLLMJSON(logsDir)
+		if err != nil {
+			b.Errorf("failed to parse vllm output: %v", err)
+		}
+
+		if metrics.Completed == 0 {
+			b.Errorf("did not complete at least one request")
+		}
+
+		for _, err := range metrics.Errors {
+			if err != "" {
+				b.Errorf("unexpected errors: %s", metrics.Errors)
+			}
+		}
+
+		b.ReportMetric(metrics.RequestsPerSec, "requests/sec")
+		b.ReportMetric(metrics.InputToksPerSec, "input-tok/sec")
+		b.ReportMetric(metrics.OutputToksPerSec, "output-tok/sec")
+		b.ReportMetric(metrics.TTFTAvgMS*1000000, "ttft-avg-ns")
+		b.ReportMetric(metrics.TTFTP50MS*1000000, "ttft-p50-ns")
+		b.ReportMetric(metrics.TTFTP99MS*1000000, "ttft-p99-ns")
+		b.ReportMetric(metrics.TPOTAvgMS*1000000, "tpot-avg-ns")
+		b.ReportMetric(metrics.TPOTP50MS*1000000, "tpot-p50-ns")
+		b.ReportMetric(metrics.TPOTP99MS*1000000, "tpot-p99-ns")
+	})
+}
+
+// Modeled after the metrics reported here: https://github.com/vllm-project/vllm/blob/main/benchmarks/benchmark_serving.py#L338-L358
+type metrics struct {
+	Duration          float64 `json:"duration"`
+	Completed         int     `json:"completed"`
+	TotalInputTokens  int     `json:"total_input_tokens"`
+	TotalOutputTokens int     `json:"total_output_tokens"`
+	RequestsPerSec    float64 `json:"request_throughput"`
+	InputToksPerSec   float64 `json:"input_throughput"`
+	OutputToksPerSec  float64 `json:"output_throughput"`
+	TTFTAvgMS         float64 `json:"mean_ttft_ms"`
+	TTFTP50MS         float64 `json:"median_ttft_ms"`
+	TTFTP99MS         float64 `json:"p99_ttft_ms"`
+	TPOTAvgMS         float64 `json:"mean_tpot_ms"`
+	TPOTP50MS         float64 `json:"median_tpot_ms"`
+	TPOTP99MS         float64 `json:"p99_tpot_ms"`
+	// metrics that are available but unused so far.
+	// InputLens         []int   `json:"input_lens"`
+	// OutputLens        []int   `json:"output_lens"`
+	// TTFTs             []float64   `json: "ttfts"`
+	// ITLS              [][]float64 `json: "itls"`
+	// GeneratedTexts    []string `json: "generated_texts"`
+	Errors []string `json: "errors"`
```

Comment:
`:vllm_test_nogo` is failing with:

```
structtag: test/gpu/vllm/vllm_test.go:150: struct field tag `json: "errors"` not compatible with reflect.StructTag.Get: bad syntax for struct tag value (GOOARCH=amd64, GOOS=linux)
```

I think the space between `:` and `"` needs to be removed.

Reply:
Aw man. Sorry about that. I tried running the test suite locally but it failed in other places. Am I able to run the CI myself to fix any remaining issues? I also couldn't find the error you posted in the [logs](https://buildkite-cloud.s3.amazonaws.com/logs-by-pipeline/b573fa4e-86a0-48f4-b8b1-769c8dc0c82b/01907a3c-502a-4314-89af-138dad32db59/01907a3c-76cd-42e7-bc6a-a1cd7ce4f45a.log?response-content-disposition=inline&response-content-type=text%2Fplain&X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=ASIAQPCP3C7LVYTQDMAM%2F20240704%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20240704T080241Z&X-Amz-Expires=600&X-Amz-Security-Token=IQoJb3JpZ2luX2VjEOj%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FwEaCXVzLWVhc3QtMSJIMEYCIQDgCZW8KBkMY1He2Lh1LGBVL7fQ%2Be7Lo9fico8XKTXnDAIhAMUr8lGiqOpUfOG4fJvQWQKRQjHSIebTsqSHvKFVOk4rKvoDCKH%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FwEQABoMMDMyMzc5NzA1MzAzIgwL2oWXsOFpXwuAtOMqzgOdlNQtVWsvThc93wuDR7%2FsayNt0DLgi62xdSLI2PR9%2FZowPvOR1ZTmL06ounf986w5dCuaffaXMX1LfWSS%2BiwLW0pNtzVoeWoGz0ErHCk3%2BI%2B0cjJ8ogO2khMI8lkGxhaOHhZczyDm6zJFocTtTW5Y7%2BGGPn2D%2F%2F3I6jI1p6sGZxZYVey1UYNAaFjwkJ9pumHBufPGc85dMfnG%2ByHQvp8dkA0AG9lTMhbCIJ3Jh%2BavSAk6WnBuIBPvHxF3wkTBGBi%2FWeEcT8DlvDCY1WTktTNAlwRzgNyR3y0ehdjyaymRv39yRdFzJLLswckmupJQCCdS7b%2BgScLSxU3qoh5vgwp1x0GPLdSbtiRKMXyFYG5WU7veLy0JZ%2F66SLT7nemwbiKK8qS8FP1Zv7swM6OSWbdNq9CkgsiVfsmitjRVY4rfS2Q7%2FlezGZ2Gq1ltDSZqtdeapUyaf%2BF2CtMA%2BvP90WfOXpYSo%2Btrlb0BsoESb6j9Tq%2B9G6zAMxE17c7LDijq2pDb06YWjCixr1EWKlk875Q%2B4wpui4G4goPhXgR1OVuyq4qPMtuYYxEfqDyM0iO7CSGFFSebSS5fABJYoRH3gwc7Po8WarM%2FMuw53mgS9jUwn6uZtAY6pAFtVfzYKjRztvyPKNLnodR7SiKxacZRWAcJ85SEjfTrmsKOV2KfdS6yDntGqYJAiJiQmU0Yv8ul%2FCeO6FTwbPvG8XAmBMVK%2FmHFzClt9C19y966ESqW93xUq4MicXYgnpPTko3hju6dGMNAowRLdrlrj89WZ8ZB0Uhk%2FWheIz4zVRkqUcqrmS2NU5Y8QTGB6joukhQJG5kdoHT%2Bj4ThiMLXiV144g%3D%3D&X-Amz-SignedHeaders=host&X-Amz-Signature=22c796a1aab55f6ba3e13c385f273ff4fdcacab279614db40975be744cc2beb5) of the failing [run](https://buildkite.com/gvisor/pipeline/builds/30575#_).

Reply:
There's another broken nogo test which was fixed on the master branch that may be overshadowing this one... Either way, it's not possible to run the BuildKite CI by yourself (because BuildKite agents run with credentials that e.g. allow them to commit to the GitHub repo so they need to run only reviewed code). But you can run the nogo tests on your own with `make test TARGETS=test/gpu/vllm/...`.